crates/optimism/flashblocks/src/service.rs

```diff
@@ -296,3 +257,64 @@ impl<
         }
     }
 }
+
+/// Simple wrapper around an ordered B-tree to keep track of a sequence of flashblocks by index
+#[derive(Debug)]
+struct FlashBlockSequence {
+    inner: BTreeMap<u64, FlashBlock>,
+}
+
+impl FlashBlockSequence {
+    const fn new() -> Self {
+        Self { inner: BTreeMap::new() }
+    }
+
+    /// Inserts a new block into the sequence.
+    ///
+    /// A [`FlashBlock`] with index 0 resets the set.
+    fn insert(&mut self, flashblock: FlashBlock) {
+        if flashblock.index == 0 {
+            trace!(number=%flashblock.block_number(), "Tracking new flashblock sequence");
+            // Flash block at index zero resets the whole state
+            self.clear();
+            self.inner.insert(flashblock.index, flashblock);
+            return
+        }
+
+        // only insert if we we previously received the same block, assume we received index 0
+        if self.block_number() == Some(flashblock.metadata.block_number) {
+            trace!(number=%flashblock.block_number(), index = %flashblock.index, block_count = self.inner.len() + 1  ,"Received followup flashblock");
+            self.inner.insert(flashblock.index, flashblock);
+        }
+    }
+
+    /// Returns the first block number
+    fn block_number(&self) -> Option<u64> {
+        Some(self.inner.values().next()?.metadata.block_number)
+    }
+
+    fn clear(&mut self) {
+        self.inner.clear();
+    }
+
+    fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Return base flashblock
+    /// Base flashblock is the first flashblock of index 0 in the sequence
+    fn base(&self) -> Option<&FlashBlock> {
+        self.inner.get(&0)
+    }
+
```

Comment:
We could also implement`parent_hash` on `FlashblockSequence` as the base is only included in the first fb, but the scope is for the whole sequence.

```suggestion

    pub fn parent_hash(&self) -> Option<B256> {
        Some(self.base()?.parent_hash)
    }

```